warning: rule starting on line 11, node 30: iov_iter_bvec(...)[1,2,12,13,16] in dtt_send_page may be inconsistently modified
warning: rule starting on line 11, node 31: sent = ... ;[1,2,12,13,17] in dtt_send_page may be inconsistently modified
warning: rule starting on line 11, node 29: bvec_set_page(...)[1,2,12,13,15] in dtt_send_page may be inconsistently modified
warning: rule starting on line 11, node 22: iov_iter_bvec(...)[1,2,9,10,13] in _dtl_send_page may be inconsistently modified
warning: rule starting on line 11, node 23: sent = ... ;[1,2,9,10,14] in _dtl_send_page may be inconsistently modified
warning: rule starting on line 11, node 21: bvec_set_page(...)[1,2,9,10,12] in _dtl_send_page may be inconsistently modified
--- drbd-headers/linux/genl_magic_struct.h
+++ /tmp/cocci-output-525084-e1a539-genl_magic_struct.h
@@ -24,6 +24,7 @@
 #define nla_strscpy nla_strlcpy
 #endif
 #include <linux/types.h>
+#include <linux/pfn_t.h>
 #include "compat.h"
 
 #define CONCAT__(a,b)	a ## b
--- drbd-headers/linux/genl_magic_func.h
+++ /tmp/cocci-output-525084-7c36f2-genl_magic_func.h
@@ -3,6 +3,7 @@
 #define GENL_MAGIC_FUNC_H
 
 #include <linux/genl_magic_struct.h>
+#include <linux/pfn_t.h>
 
 /*
  * Magic: declare tla policy						{{{1
--- drbd-headers/linux/drbd_genl_api.h
+++ /tmp/cocci-output-525084-ae03b9-drbd_genl_api.h
@@ -39,5 +39,6 @@ enum {
 #define GENL_MAGIC_FAMILY_HDRSZ	sizeof(struct drbd_genlmsghdr)
 #define GENL_MAGIC_INCLUDE_FILE <linux/drbd_genl.h>
 #include <linux/genl_magic_struct.h>
+#include <linux/pfn_t.h>
 
 #endif
--- drbd-headers/linux/drbd.h
+++ /tmp/cocci-output-525084-f35a76-drbd.h
@@ -22,6 +22,7 @@
 #include <sys/types.h>
 #include <sys/wait.h>
 #include <limits.h>
+#include <linux/pfn_t.h>
 
 /* Although the Linux source code makes a difference between
    generic endianness and the bitfields' endianness, there is no
--- drbd-kernel-compat/handshake/tlshd.c
+++ /tmp/cocci-output-525084-221c06-tlshd.c
@@ -21,6 +21,7 @@
 
 #include <uapi/linux/keyctl.h>
 #include <uapi/linux/handshake.h>
+#include <linux/pfn_t.h>
 #include "handshake.h"
 
 struct tls_handshake_req {
--- drbd-kernel-compat/handshake/request.c
+++ /tmp/cocci-output-525084-5db7e6-request.c
@@ -21,6 +21,7 @@
 #include <net/netns/generic.h>
 
 #include <uapi/linux/handshake.h>
+#include <linux/pfn_t.h>
 #include "handshake.h"
 
 /*
--- drbd-kernel-compat/handshake/netlink.c
+++ /tmp/cocci-output-525084-aab659-netlink.c
@@ -21,6 +21,7 @@
 #include <net/netns/generic.h>
 
 #include <uapi/linux/handshake.h>
+#include <linux/pfn_t.h>
 #include "handshake.h"
 #include "genl.h"
 
--- drbd-kernel-compat/handshake/genl.c
+++ /tmp/cocci-output-525084-cbb6dc-genl.c
@@ -9,6 +9,7 @@
 #include "genl.h"
 
 #include <uapi/linux/handshake.h>
+#include <linux/pfn_t.h>
 
 
 /* HANDSHAKE_CMD_ACCEPT - do */
--- kref_debug.c
+++ /tmp/cocci-output-525084-5eaddd-kref_debug.c
@@ -4,6 +4,7 @@
 #include <linux/spinlock.h>
 #include <linux/seq_file.h>
 #include <linux/kref.h>
+#include <linux/pfn_t.h>
 #include "kref_debug.h"
 
 static struct list_head kref_debug_objects;
--- drbd_transport_template.c
+++ /tmp/cocci-output-525084-333853-drbd_transport_template.c
@@ -1,5 +1,6 @@
 // SPDX-License-Identifier: GPL-2.0-only
 #include <linux/module.h>
+#include <linux/pfn_t.h>
 #include "drbd_transport.h"
 #include "drbd_int.h"
 
--- drbd_transport_tcp.c
+++ /tmp/cocci-output-525084-1d2444-drbd_transport_tcp.c
@@ -26,6 +26,7 @@
 #include <net/handshake.h>
 #include <net/tls.h>
 #include <net/tls_prot.h>
+#include <linux/pfn_t.h>
 #include "drbd_protocol.h"
 #include "drbd_transport.h"
 
@@ -520,15 +521,15 @@ static int dtt_try_connect(struct dtt_pa
 	*  a free one dynamically.
 	*/
 	what = "bind before connect";
-	err = socket->ops->bind(socket, (struct sockaddr_unsized *) &my_addr,
-			path->path.my_addr_len);
+	err = socket->ops->bind(socket, (struct sockaddr *) &my_addr,
+				path->path.my_addr_len);
 	if (err < 0)
 		goto out;
 
 	/* connect may fail, peer not yet available.
 	 * stay C_CONNECTING, don't go Disconnecting! */
 	what = "connect";
-	err = socket->ops->connect(socket, (struct sockaddr_unsized *) &peer_addr,
+	err = socket->ops->connect(socket, (struct sockaddr *) &peer_addr,
 				   path->path.peer_addr_len, 0);
 	if (err < 0) {
 		err = dtt_socket_err_to_eagain(err);
@@ -828,7 +829,7 @@ retry:
 				container_of(drbd_path2->transport,
 					     struct drbd_tcp_transport, transport);
 
-			socket_c = kmalloc_obj(*socket_c, GFP_ATOMIC);
+			socket_c = kmalloc(sizeof(*socket_c), GFP_ATOMIC);
 			if (!socket_c) {
 				tr_info(transport, /* path2->transport, */
 					"No mem, dropped an incoming connection\n");
@@ -1019,7 +1020,7 @@ static void dtt_incoming_connection(stru
 
 static void dtt_control_timer_fn(struct timer_list *t)
 {
-	struct drbd_tcp_transport *tcp_transport = timer_container_of(tcp_transport, t,
+	struct drbd_tcp_transport *tcp_transport = from_timer(tcp_transport, t,
 			control_timer);
 	struct drbd_transport *transport = &tcp_transport->transport;
 
@@ -1074,7 +1075,8 @@ static int dtt_init_listener(struct drbd
 	addr_len = addr->sa_family == AF_INET6 ? sizeof(struct sockaddr_in6)
 		: sizeof(struct sockaddr_in);
 
-	err = s_listen->ops->bind(s_listen, (struct sockaddr_unsized *)&my_addr, addr_len);
+	err = s_listen->ops->bind(s_listen, (struct sockaddr *)&my_addr,
+				  addr_len);
 	if (err < 0) {
 		what = "bind before listen";
 		goto out;
@@ -1537,8 +1539,7 @@ static int dtt_send_page(struct drbd_tra
 	struct drbd_tcp_transport *tcp_transport =
 		container_of(transport, struct drbd_tcp_transport, transport);
 	struct socket *socket = tcp_transport->stream[stream];
-	struct msghdr msg = { .msg_flags = msg_flags | MSG_NOSIGNAL | MSG_SPLICE_PAGES };
-	struct bio_vec bvec;
+	struct msghdr msg = { .msg_flags = msg_flags | MSG_NOSIGNAL };
 	int len = size;
 	int err = -EIO;
 
@@ -1549,10 +1550,8 @@ static int dtt_send_page(struct drbd_tra
 	do {
 		int sent;
 
-		bvec_set_page(&bvec, page, len, offset);
-		iov_iter_bvec(&msg.msg_iter, ITER_SOURCE, &bvec, 1, len);
-
-		sent = sock_sendmsg(socket, &msg);
+		sent = socket->ops->sendpage(socket, page, offset, len,
+					     msg.msg_flags);
 		if (sent <= 0) {
 			if (sent == -EAGAIN) {
 				if (drbd_stream_send_timed_out(transport, stream))
--- drbd_transport_rdma.c
+++ /tmp/cocci-output-525084-8480df-drbd_transport_rdma.c
@@ -29,6 +29,7 @@
 #include <rdma/ib_cm.h>
 #include <linux/interrupt.h>
 #include <linux/drbd_genl_api.h>
+#include <linux/pfn_t.h>
 #include "drbd_protocol.h"
 #include "drbd_transport.h"
 #include "linux/drbd_config.h" /* for REL_VERSION */
@@ -536,7 +537,7 @@ static void dtr_free(struct drbd_transpo
 
 static void dtr_control_timer_fn(struct timer_list *t)
 {
-	struct dtr_transport *rdma_transport = timer_container_of(rdma_transport, t, control_timer);
+	struct dtr_transport *rdma_transport = from_timer(rdma_transport, t, control_timer);
 	struct drbd_transport *transport = &rdma_transport->transport;
 
 	drbd_control_event(transport, TIMEOUT);
@@ -1629,7 +1630,7 @@ out:
 
 static void dtr_tx_timeout_fn(struct timer_list *t)
 {
-	struct dtr_cm *cm = timer_container_of(cm, t, tx_timeout);
+	struct dtr_cm *cm = from_timer(cm, t, tx_timeout);
 
 	/* cm->kref for armed timer becomes a ref for the work */
 	schedule_work(&cm->tx_timeout_work);
--- drbd_transport_lb-tcp.c
+++ /tmp/cocci-output-525084-10ab6f-drbd_transport_lb-tcp.c
@@ -18,6 +18,7 @@
 #include <linux/drbd_genl_api.h>
 #include <linux/drbd_config.h>
 #include <net/tcp.h>
+#include <linux/pfn_t.h>
 #include "drbd_protocol.h"
 #include "drbd_transport.h"
 
@@ -628,7 +629,8 @@ dtl_try_connect(struct drbd_transport *t
 	 * a free one dynamically.
 	 */
 	what = "bind before connect";
-	err = sock->ops->bind(sock, (struct sockaddr_unsized *) &my_addr, path->path.my_addr_len);
+	err = sock->ops->bind(sock, (struct sockaddr *) &my_addr,
+			      path->path.my_addr_len);
 	if (err < 0)
 		goto out;
 
@@ -637,7 +639,7 @@ dtl_try_connect(struct drbd_transport *t
 	mutex_lock(&dtl_transport->connecting_socket_mutex);
 	dtl_transport->connecting_socket = sock;
 	mutex_unlock(&dtl_transport->connecting_socket_mutex);
-	err = sock->ops->connect(sock, (struct sockaddr_unsized *) &peer_addr,
+	err = sock->ops->connect(sock, (struct sockaddr *) &peer_addr,
 				   path->path.peer_addr_len, 0);
 	mutex_lock(&dtl_transport->connecting_socket_mutex);
 	dtl_transport->connecting_socket = NULL;
@@ -1033,7 +1035,7 @@ static void dtl_incoming_connection(stru
 
 static void dtl_control_timer_fn(struct timer_list *t)
 {
-	struct dtl_transport *dtl_transport = timer_container_of(dtl_transport, t, control_timer);
+	struct dtl_transport *dtl_transport = from_timer(dtl_transport, t, control_timer);
 	struct drbd_transport *transport = &dtl_transport->transport;
 
 	drbd_control_event(transport, TIMEOUT);
@@ -1089,7 +1091,8 @@ static int dtl_init_listener(struct drbd
 	addr_len = addr->sa_family == AF_INET6 ? sizeof(struct sockaddr_in6)
 		: sizeof(struct sockaddr_in);
 
-	err = s_listen->ops->bind(s_listen, (struct sockaddr_unsized *)&my_addr, addr_len);
+	err = s_listen->ops->bind(s_listen, (struct sockaddr *)&my_addr,
+				  addr_len);
 	if (err < 0) {
 		what = "bind before listen";
 		goto out;
@@ -1660,20 +1663,17 @@ static int dtl_select_send_flow(struct d
 static int _dtl_send_page(struct dtl_transport *dtl_transport, struct dtl_flow *flow,
 			  struct page *page, int offset, size_t size, unsigned int msg_flags)
 {
-	struct msghdr msg = { .msg_flags = msg_flags | MSG_NOSIGNAL | MSG_SPLICE_PAGES };
+	struct msghdr msg = { .msg_flags = msg_flags | MSG_NOSIGNAL };
 	struct drbd_transport *transport = &dtl_transport->transport;
 	struct socket *sock = flow->sock;
-	struct bio_vec bvec;
 	int len = size;
 	int err = -EIO;
 
 	do {
 		int sent;
 
-		bvec_set_page(&bvec, page, len, offset);
-		iov_iter_bvec(&msg.msg_iter, ITER_SOURCE, &bvec, 1, len);
-
-		sent = sock_sendmsg(sock, &msg);
+		sent = sock->ops->sendpage(sock, page, offset, len,
+					   msg.msg_flags);
 		if (sent <= 0) {
 			if (sent == -EAGAIN) {
 				if (drbd_stream_send_timed_out(transport, flow->stream_nr))
--- drbd_transport.c
+++ /tmp/cocci-output-525084-e69b97-drbd_transport.c
@@ -4,6 +4,7 @@
 #include <linux/spinlock.h>
 #include <linux/module.h>
 #include <net/ipv6.h>
+#include <linux/pfn_t.h>
 #include "drbd_transport.h"
 #include "drbd_int.h"
 
--- drbd_state.c
+++ /tmp/cocci-output-525084-9a33b3-drbd_state.c
@@ -15,6 +15,7 @@
 
 #include <linux/random.h>
 #include <linux/jiffies.h>
+#include <linux/pfn_t.h>
 #include "drbd_int.h"
 #include "drbd_protocol.h"
 #include "drbd_req.h"
@@ -791,7 +792,7 @@ static struct after_state_change_work *a
 	if (test_bit(R_UNREGISTERED, &resource->flags))
 		return NULL;
 
-	work = kmalloc_obj(*work, GFP_ATOMIC);
+	work = kmalloc(sizeof(*work), GFP_ATOMIC);
 	if (work) {
 		work->state_change = remember_state_change(resource, GFP_ATOMIC);
 		if (!work->state_change) {
--- drbd_sender.c
+++ /tmp/cocci-output-525084-000611-drbd_sender.c
@@ -22,6 +22,7 @@
 #include <linux/random.h>
 #include <linux/overflow.h>
 #include <linux/part_stat.h>
+#include <linux/pfn_t.h>
 
 #include "drbd_int.h"
 #include "drbd_protocol.h"
@@ -774,7 +775,7 @@ bool drbd_any_flush_pending(struct drbd_
 
 void resync_timer_fn(struct timer_list *t)
 {
-	struct drbd_peer_device *peer_device = timer_container_of(peer_device, t, resync_timer);
+	struct drbd_peer_device *peer_device = from_timer(peer_device, t, resync_timer);
 
 	drbd_queue_work_if_unqueued(
 		&peer_device->connection->sender_work,
@@ -814,7 +815,7 @@ struct fifo_buffer *fifo_alloc(unsigned
 {
 	struct fifo_buffer *fb;
 
-	fb = kzalloc_flex(*fb, values, fifo_size, GFP_NOIO);
+	fb = kzalloc(struct_size(fb, values, fifo_size), GFP_NOIO);
 	if (!fb)
 		return NULL;
 
@@ -1759,7 +1760,7 @@ static void queue_resync_finished(struct
 	struct drbd_connection *connection = peer_device->connection;
 	struct resync_finished_work *rfw;
 
-	rfw = kmalloc_obj(*rfw, GFP_ATOMIC);
+	rfw = kmalloc(sizeof(*rfw), GFP_ATOMIC);
 	if (!rfw) {
 		drbd_err(peer_device, "Warn failed to kmalloc(dw).\n");
 		return;
@@ -2916,7 +2917,7 @@ void drbd_rs_controller_reset(struct drb
 
 void start_resync_timer_fn(struct timer_list *t)
 {
-	struct drbd_peer_device *peer_device = timer_container_of(peer_device, t,
+	struct drbd_peer_device *peer_device = from_timer(peer_device, t,
 			start_resync_timer);
 	drbd_peer_device_post_work(peer_device, RS_START);
 }
--- drbd_req.c
+++ /tmp/cocci-output-525084-bd4062-drbd_req.c
@@ -15,6 +15,7 @@
 
 #include <linux/slab.h>
 #include <linux/drbd.h>
+#include <linux/pfn_t.h>
 #include "drbd_int.h"
 #include "drbd_req.h"
 
@@ -140,7 +141,7 @@ int w_queue_peer_ack(struct drbd_work *w
 
 	list_for_each_entry_safe(req, tmp, &work_list, list) {
 		struct drbd_peer_ack *peer_ack =
-			kzalloc_obj(struct drbd_peer_ack);
+			kzalloc(sizeof(struct drbd_peer_ack), GFP_KERNEL);
 
 		peer_ack->resource = resource;
 		INIT_LIST_HEAD(&peer_ack->list);
@@ -2876,7 +2877,7 @@ static bool net_timeout_reached(struct d
 
 void request_timer_fn(struct timer_list *t)
 {
-	struct drbd_device *device = timer_container_of(device, t, request_timer);
+	struct drbd_device *device = from_timer(device, t, request_timer);
 	struct drbd_resource *resource = device->resource;
 	struct drbd_connection *connection;
 	struct drbd_request *req_read, *req_write;
--- drbd_receiver.c
+++ /tmp/cocci-output-525084-e4f041-drbd_receiver.c
@@ -27,6 +27,7 @@
 #include <linux/random.h>
 #include <net/ipv6.h>
 #include <linux/part_stat.h>
+#include <linux/pfn_t.h>
 
 #include "drbd_int.h"
 #include "drbd_protocol.h"
@@ -733,7 +734,7 @@ void wait_initial_states_received(struct
 
 void connect_timer_fn(struct timer_list *t)
 {
-	struct drbd_connection *connection = timer_container_of(connection, t, connect_timer);
+	struct drbd_connection *connection = from_timer(connection, t, connect_timer);
 
 	drbd_queue_work(&connection->sender_work, &connection->connect_timer_work);
 }
@@ -1261,7 +1262,7 @@ static void submit_one_flush(struct drbd
 {
 	struct bio *bio = bio_alloc(device->ldev->backing_bdev, 0,
 			REQ_OP_WRITE | REQ_PREFLUSH, GFP_NOIO);
-	struct one_flush_context *octx = kmalloc_obj(*octx, GFP_NOIO);
+	struct one_flush_context *octx = kmalloc(sizeof(*octx), GFP_NOIO);
 
 	if (!octx) {
 		drbd_warn(device, "Could not allocate a octx, CANNOT ISSUE FLUSH\n");
@@ -1514,7 +1515,7 @@ static enum finish_epoch drbd_may_finish
 
 	if (schedule_flush) {
 		struct flush_work *fw;
-		fw = kmalloc_obj(*fw, GFP_ATOMIC);
+		fw = kmalloc(sizeof(*fw), GFP_ATOMIC);
 		if (fw) {
 			fw->w.cb = w_flush;
 			fw->epoch = epoch;
@@ -2015,7 +2016,7 @@ static int receive_Barrier(struct drbd_c
 
 	/* receiver context, in the writeout path of the other node.
 	 * avoid potential distributed deadlock */
-	epoch = kzalloc_obj(struct drbd_epoch, GFP_NOIO);
+	epoch = kzalloc(sizeof(struct drbd_epoch), GFP_NOIO);
 	if (!epoch) {
 		drbd_warn(connection, "Allocation of an epoch failed, slowing down\n");
 		issue_flush = !test_and_set_bit(DE_BARRIER_IN_NEXT_EPOCH_ISSUED, &connection->current_epoch->flags);
@@ -2353,7 +2354,7 @@ void drbd_queue_update_peers(struct drbd
 	struct drbd_device *device = peer_device->device;
 	struct update_peers_work *upw;
 
-	upw = kmalloc_obj(*upw, GFP_ATOMIC | __GFP_NOWARN);
+	upw = kmalloc(sizeof(*upw), GFP_ATOMIC | __GFP_NOWARN);
 	if (upw) {
 		upw->sector_start = sector_start;
 		upw->sector_end = sector_end;
@@ -5780,7 +5781,7 @@ static int receive_protocol(struct drbd_
 		}
 	}
 
-	new_net_conf = kmalloc_obj(struct net_conf);
+	new_net_conf = kmalloc(sizeof(struct net_conf), GFP_KERNEL);
 	if (!new_net_conf)
 		goto disconnect;
 
@@ -5977,7 +5978,8 @@ static int receive_SyncParam(struct drbd
 		}
 
 		if (verify_tfm || csums_tfm) {
-			new_net_conf = kzalloc_obj(struct net_conf);
+			new_net_conf = kzalloc(sizeof(struct net_conf),
+					       GFP_KERNEL);
 			if (!new_net_conf)
 				goto disconnect;
 
@@ -6260,7 +6262,8 @@ static int receive_sizes(struct drbd_con
 		if (my_usize != p_usize) {
 			struct disk_conf *old_disk_conf, *new_disk_conf;
 
-			new_disk_conf = kzalloc_obj(struct disk_conf);
+			new_disk_conf = kzalloc(sizeof(struct disk_conf),
+						GFP_KERNEL);
 			if (!new_disk_conf) {
 				err = -ENOMEM;
 				goto out;
@@ -7257,7 +7260,7 @@ static void drbd_abort_twopc(struct drbd
 
 void twopc_timer_fn(struct timer_list *t)
 {
-	struct drbd_resource *resource = timer_container_of(resource, t, twopc_timer);
+	struct drbd_resource *resource = from_timer(resource, t, twopc_timer);
 	unsigned long irq_flags;
 
 	write_lock_irqsave(&resource->state_rwlock, irq_flags);
@@ -7592,7 +7595,7 @@ drbd_commit_size_change(struct drbd_devi
 		drbd_info(device, "New u_size %llu sectors\n",
 			  (unsigned long long)tr->user_size);
 
-		new_disk_conf = kzalloc_obj(struct disk_conf);
+		new_disk_conf = kzalloc(sizeof(struct disk_conf), GFP_KERNEL);
 		if (!new_disk_conf) {
 			device->ldev->disk_conf->disk_size = tr->user_size;
 			goto cont;
@@ -10715,7 +10718,7 @@ int drbd_do_auth(struct drbd_connection
 		goto fail;
 	}
 
-	peers_ch = kmalloc_obj(*peers_ch, GFP_NOIO);
+	peers_ch = kmalloc(sizeof(*peers_ch), GFP_NOIO);
 	if (!peers_ch) {
 		rv = -1;
 		goto fail;
--- drbd_proc.c
+++ /tmp/cocci-output-525084-18c4aa-drbd_proc.c
@@ -13,6 +13,7 @@
 
 #include <linux/proc_fs.h>
 #include <linux/seq_file.h>
+#include <linux/pfn_t.h>
 #include "drbd_int.h"
 #include "drbd_transport.h"
 #include "drbd_legacy_84.h"
--- drbd_nla.c
+++ /tmp/cocci-output-525084-075845-drbd_nla.c
@@ -2,6 +2,7 @@
 #include <linux/kernel.h>
 #include <net/netlink.h>
 #include <linux/drbd_genl_api.h>
+#include <linux/pfn_t.h>
 #include "drbd_nla.h"
 
 static int drbd_nla_check_mandatory(int maxtype, struct nlattr *nla)
--- drbd_nl.c
+++ /tmp/cocci-output-525084-7fb58f-drbd_nl.c
@@ -86,6 +86,7 @@ static int drbd_adm_get_initial_state_do
 #include <linux/drbd_genl_api.h>
 #include "drbd_nla.h"
 #include <linux/genl_magic_func.h>
+#include <linux/pfn_t.h>
 
 void drbd_enable_netns(void)
 {
@@ -2089,7 +2090,7 @@ static void get_common_queue_limits(stru
 	common_limits->physical_block_size = device->device_conf.block_size;
 	common_limits->logical_block_size = device->device_conf.block_size;
 	common_limits->io_min = device->device_conf.block_size;
-	common_limits->max_hw_zone_append_sectors = 0;
+	common_limits->max_zone_append_sectors = 0;
 
 	rcu_read_lock();
 	for_each_peer_device_rcu(peer_device, device) {
@@ -2115,7 +2116,7 @@ void drbd_reconsider_queue_parameters(st
 	struct queue_limits lim;
 	struct request_queue *b = NULL;
 
-	lim = queue_limits_start_update(q);
+	lim = q->limits;
 	get_common_queue_limits(&lim, device);
 
 	/*
@@ -2136,24 +2137,7 @@ void drbd_reconsider_queue_parameters(st
 	if (bdev) {
 		b = bdev->backing_bdev->bd_disk->queue;
 		blk_stack_limits(&lim, &b->limits, 0);
-		/*
-		 * blk_set_stacking_limits() cleared the features, and
-		 * blk_stack_limits() may or may not have inherited
-		 * BLK_FEAT_STABLE_WRITES from the backing device.
-		 *
-		 * DRBD always requires stable writes because:
-		 * 1. The same bio data is read for both local disk I/O and
-		 *    network transmission. If the page changes mid-flight,
-		 *    the local and remote copies could diverge.
-		 * 2. When data integrity is enabled, DRBD calculates a
-		 *    checksum before sending the data. If the page changes
-		 *    between checksum calculation and transmission, the
-		 *    receiver will detect a checksum mismatch.
-		 */
-		lim.features |= BLK_FEAT_STABLE_WRITES;
 	} else {
-		lim.features = BLK_FEAT_WRITE_CACHE | BLK_FEAT_FUA |
-			       BLK_FEAT_ROTATIONAL | BLK_FEAT_STABLE_WRITES;
 	}
 
 	/*
@@ -2165,7 +2149,6 @@ void drbd_reconsider_queue_parameters(st
 		lim.max_write_zeroes_sectors = DRBD_MAX_BBIO_SECTORS;
 	else
 		lim.max_write_zeroes_sectors = 0;
-	lim.max_hw_wzeroes_unmap_sectors = 0;
 
 	if ((lim.discard_granularity >> SECTOR_SHIFT) >
 	    lim.max_hw_discard_sectors) {
@@ -2178,8 +2161,11 @@ void drbd_reconsider_queue_parameters(st
 		lim.max_hw_discard_sectors = 0;
 	}
 
-	if (queue_limits_commit_update(q, &lim))
-		drbd_err(device, "setting new queue limits failed\n");
+	q->limits = lim;
+	blk_queue_max_hw_sectors(q, lim.max_hw_sectors);
+	blk_queue_max_discard_sectors(q, lim.max_hw_discard_sectors);
+	if (bdev)
+		disk_update_readahead(device->vdisk);
 }
 
 /* Make sure IO is suspended before calling this function(). */
@@ -2459,7 +2445,7 @@ static int drbd_adm_disk_opts(struct sk_
 		goto out;
 	}
 
-	new_disk_conf = kmalloc_obj(struct disk_conf);
+	new_disk_conf = kmalloc(sizeof(struct disk_conf), GFP_KERNEL);
 	if (!new_disk_conf) {
 		retcode = ERR_NOMEM;
 		goto fail;
@@ -3229,14 +3215,14 @@ static int drbd_adm_attach(struct sk_buf
 	}
 
 	/* allocation not in the IO path, drbdsetup context */
-	nbc = kzalloc_obj(struct drbd_backing_dev);
+	nbc = kzalloc(sizeof(struct drbd_backing_dev), GFP_KERNEL);
 	if (!nbc) {
 		retcode = ERR_NOMEM;
 		goto fail;
 	}
 	spin_lock_init(&nbc->md.uuid_lock);
 
-	new_disk_conf = kzalloc_obj(struct disk_conf);
+	new_disk_conf = kzalloc(sizeof(struct disk_conf), GFP_KERNEL);
 	if (!new_disk_conf) {
 		retcode = ERR_NOMEM;
 		goto fail;
@@ -4095,7 +4081,7 @@ static int drbd_adm_net_opts(struct sk_b
 		goto out_no_adm_mutex;
 	}
 
-	new_net_conf = kzalloc_obj(struct net_conf);
+	new_net_conf = kzalloc(sizeof(struct net_conf), GFP_KERNEL);
 	if (!new_net_conf) {
 		retcode = ERR_NOMEM;
 		goto out;
@@ -4261,7 +4247,8 @@ static int drbd_adm_peer_device_opts(str
 	}
 	mutex_lock(&adm_ctx.resource->conf_update);
 
-	new_peer_device_conf = kzalloc_obj(struct peer_device_conf);
+	new_peer_device_conf = kzalloc(sizeof(struct peer_device_conf),
+				       GFP_KERNEL);
 	if (!new_peer_device_conf)
 		goto fail;
 
@@ -4352,7 +4339,7 @@ int drbd_create_peer_device_default_conf
 	struct peer_device_conf *conf;
 	int err;
 
-	conf = kzalloc_obj(*conf);
+	conf = kzalloc(sizeof(*conf), GFP_KERNEL);
 	if (!conf)
 		return -ENOMEM;
 
@@ -4505,7 +4492,7 @@ static int adm_new_connection(struct drb
 	struct drbd_transport *transport;
 
 	/* allocation not in the IO path, drbdsetup / netlink process context */
-	new_net_conf = kzalloc_obj(*new_net_conf);
+	new_net_conf = kzalloc(sizeof(*new_net_conf), GFP_KERNEL);
 	if (!new_net_conf)
 		return ERR_NOMEM;
 
@@ -5288,7 +5275,7 @@ sector_t drbd_local_max_size(struct drbd
 	struct drbd_backing_dev *tmp_bdev;
 	sector_t s;
 
-	tmp_bdev = kmalloc_obj(struct drbd_backing_dev, GFP_ATOMIC);
+	tmp_bdev = kmalloc(sizeof(struct drbd_backing_dev), GFP_ATOMIC);
 	if (!tmp_bdev)
 		return 0;
 
@@ -5401,7 +5388,7 @@ static int drbd_adm_resize(struct sk_buf
 	u_size = rcu_dereference(device->ldev->disk_conf)->disk_size;
 	rcu_read_unlock();
 	if (u_size != (sector_t)rs.resize_size) {
-		new_disk_conf = kmalloc_obj(struct disk_conf);
+		new_disk_conf = kmalloc(sizeof(struct disk_conf), GFP_KERNEL);
 		if (!new_disk_conf) {
 			retcode = ERR_NOMEM;
 			goto fail_ldev;
--- drbd_main.c
+++ /tmp/cocci-output-525084-add7f4-drbd_main.c
@@ -71,6 +71,7 @@ MODULE_ALIAS_BLOCKDEV_MAJOR(DRBD_MAJOR);
 MODULE_SOFTDEP("post: handshake");
 
 #include <linux/moduleparam.h>
+#include <linux/pfn_t.h>
 
 #ifdef CONFIG_DRBD_FAULT_INJECTION
 int drbd_enable_faults;
@@ -2846,7 +2847,7 @@ static void add_opener(struct drbd_devic
 		resource->auto_promoted_by.opened = now;
 		get_task_comm(resource->auto_promoted_by.comm, current);
 	}
-	opener = kmalloc_obj(*opener, GFP_NOIO);
+	opener = kmalloc(sizeof(*opener), GFP_NOIO);
 	if (!opener)
 		return;
 	get_task_comm(opener->comm, current);
@@ -3675,7 +3676,7 @@ void drbd_flush_peer_acks(struct drbd_re
 
 static void peer_ack_timer_fn(struct timer_list *t)
 {
-	struct drbd_resource *resource = timer_container_of(resource, t, peer_ack_timer);
+	struct drbd_resource *resource = from_timer(resource, t, peer_ack_timer);
 
 	drbd_flush_peer_acks(resource);
 }
@@ -3790,7 +3791,7 @@ struct drbd_resource *drbd_create_resour
 {
 	struct drbd_resource *resource;
 
-	resource = kzalloc_obj(struct drbd_resource);
+	resource = kzalloc(sizeof(struct drbd_resource), GFP_KERNEL);
 	if (!resource)
 		goto fail;
 	resource->name = kstrdup(name, GFP_KERNEL);
@@ -3875,7 +3876,8 @@ struct drbd_connection *drbd_create_conn
 	if (drbd_alloc_send_buffers(connection))
 		goto fail;
 
-	connection->current_epoch = kzalloc_obj(struct drbd_epoch);
+	connection->current_epoch = kzalloc(sizeof(struct drbd_epoch),
+					    GFP_KERNEL);
 	if (!connection->current_epoch)
 		goto fail;
 
@@ -4039,7 +4041,7 @@ struct drbd_peer_device *create_peer_dev
 	struct drbd_peer_device *peer_device;
 	int err;
 
-	peer_device = kzalloc_obj(struct drbd_peer_device);
+	peer_device = kzalloc(sizeof(struct drbd_peer_device), GFP_KERNEL);
 	if (!peer_device)
 		return NULL;
 
@@ -4158,7 +4160,7 @@ enum drbd_ret_code drbd_create_device(st
 		return ERR_MINOR_OR_VOLUME_EXISTS;
 
 	/* GFP_KERNEL, we are outside of all write-out paths */
-	device = kzalloc_obj(struct drbd_device);
+	device = kzalloc(sizeof(struct drbd_device), GFP_KERNEL);
 	if (!device)
 		return ERR_NOMEM;
 	kref_init(&device->kref);
@@ -4212,9 +4214,8 @@ enum drbd_ret_code drbd_create_device(st
 
 	init_rwsem(&device->uuid_sem);
 
-	disk = blk_alloc_disk(NULL, NUMA_NO_NODE);
-	if (IS_ERR(disk)) {
-		err = PTR_ERR(disk);
+	disk = blk_alloc_disk(NUMA_NO_NODE);
+	if (!disk) {
 		goto out_no_disk;
 	}
 
@@ -4230,6 +4231,8 @@ enum drbd_ret_code drbd_create_device(st
 	disk->flags |= GENHD_FL_NO_PART;
 	sprintf(disk->disk_name, "drbd%d", minor);
 	disk->private_data = device;
+	blk_queue_flag_set(QUEUE_FLAG_STABLE_WRITES, disk->queue);
+	blk_queue_write_cache(disk->queue, true, true);
 
 	device->md_io.page = alloc_page(GFP_KERNEL);
 	if (!device->md_io.page)
@@ -6122,7 +6125,7 @@ void drbd_queue_bitmap_io(struct drbd_de
 
 	D_ASSERT(device, current == device->resource->worker.task);
 
-	bm_io_work = kmalloc_obj(*bm_io_work, GFP_NOIO);
+	bm_io_work = kmalloc(sizeof(*bm_io_work), GFP_NOIO);
 	if (!bm_io_work) {
 		if (done)
 			done(device, peer_device, -ENOMEM);
@@ -6257,7 +6260,7 @@ bool drbd_md_test_peer_flag(struct drbd_
 
 static void md_sync_timer_fn(struct timer_list *t)
 {
-	struct drbd_device *device = timer_container_of(device, t, md_sync_timer);
+	struct drbd_device *device = from_timer(device, t, md_sync_timer);
 	drbd_device_post_work(device, MD_SYNC);
 }
 
--- drbd_kref_debug.c
+++ /tmp/cocci-output-525084-323cbe-drbd_kref_debug.c
@@ -1,5 +1,6 @@
 // SPDX-License-Identifier: GPL-2.0-only
 #include <drbd_kref_debug.h>
+#include <linux/pfn_t.h>
 #include "drbd_int.h"
 
 
--- drbd_interval.c
+++ /tmp/cocci-output-525084-56fefb-drbd_interval.c
@@ -1,6 +1,7 @@
 // SPDX-License-Identifier: GPL-2.0-only
 #include <asm/bug.h>
 #include <linux/rbtree_augmented.h>
+#include <linux/pfn_t.h>
 #include "drbd_interval.h"
 
 /*
--- drbd_debugfs.c
+++ /tmp/cocci-output-525084-524549-drbd_debugfs.c
@@ -7,6 +7,7 @@
 #include <linux/jiffies.h>
 #include <linux/list.h>
 #include <generated/utsrelease.h>
+#include <linux/pfn_t.h>
 
 #include "drbd_int.h"
 #include "drbd_req.h"
@@ -778,11 +779,16 @@ void drbd_debugfs_resource_cleanup(struc
 
 void drbd_debugfs_resource_rename(struct drbd_resource *resource, const char *new_name)
 {
-	int err;
-
-	err = debugfs_change_name(resource->debugfs_res, "%s", new_name);
-	if (err)
+	struct dentry *new_d;new_d = debugfs_rename(drbd_debugfs_resources,
+						    resource->debugfs_res,
+						    drbd_debugfs_resources,
+						    new_name);
+	if (IS_ERR(new_d)) {
 		drbd_err(resource, "failed to rename debugfs entry for resource\n");
+	} else
+	{
+		resource->debugfs_res = new_d;
+	}
 }
 
 static void seq_print_one_timing_detail(struct seq_file *m,
@@ -1964,6 +1970,22 @@ static const struct file_operations drbd
 
 static int drbd_compat_show(struct seq_file *m, void *ignored)
 {
+	seq_puts(m, "sendpage__yes_present\n");
+	seq_puts(m, "queue_limits_features__no_present\n");
+	seq_puts(m, "blk_alloc_disk__no_takes_queue_limits\n");
+	seq_puts(m, "alloc_obj_default_gfp__no_present\n");
+	seq_puts(m, "kmalloc_obj__no_present\n");
+	seq_puts(m, "kzalloc_flex__no_present\n");
+	seq_puts(m, "queue_limits_start_update__no_present\n");
+	seq_puts(m, "blk_queue_update_readahead__no_present\n");
+	seq_puts(m, "pfn_t__yes_present\n");
+	seq_puts(m, "linux_unaligned_h__no_present\n");
+	seq_puts(m, "genlmsg_multicast_allns__no_has_4_params\n");
+	seq_puts(m, "debugfs_change_name__no_present\n");
+	seq_puts(m, "queue_limits_max_hw_zone_append_sectors__no_present\n");
+	seq_puts(m, "timer_container_of__no_present\n");
+	seq_puts(m, "queue_limits_max_hw_wzeroes_unmap_sectors__no_present\n");
+	seq_puts(m, "struct_sockaddr_unsized__no_present\n");
 	return 0;
 }
 
--- drbd_dax_pmem.c
+++ /tmp/cocci-output-525084-3026b0-drbd_dax_pmem.c
@@ -24,6 +24,7 @@
 #include <linux/dax.h>
 #include <linux/libnvdimm.h>
 #include <linux/blkdev.h>
+#include <linux/pfn_t.h>
 #include "drbd_int.h"
 #include "drbd_dax_pmem.h"
 #include "drbd_meta_data.h"
@@ -34,7 +35,7 @@ static int map_superblock_for_dax(struct
 	pgoff_t pgoff = bdev->md.md_offset >> (PAGE_SHIFT - SECTOR_SHIFT);
 	void *kaddr;
 	long len;
-	unsigned long pfn_unused; /* before 4.18 it is required to pass in non-NULL */
+	pfn_t pfn_unused; /* before 4.18 it is required to pass in non-NULL */
 	int id;
 
 	id = dax_read_lock();
@@ -92,7 +93,7 @@ int drbd_dax_map(struct drbd_backing_dev
 	long al_offset_byte = (al_sector - first_sector) << SECTOR_SHIFT;
 	void *kaddr;
 	long len;
-	unsigned long pfn_unused; /* before 4.18 it is required to pass in non-NULL */
+	pfn_t pfn_unused; /* before 4.18 it is required to pass in non-NULL */
 	int id;
 
 	id = dax_read_lock();
--- drbd_bitmap.c
+++ /tmp/cocci-output-525084-bde8b2-drbd_bitmap.c
@@ -19,6 +19,7 @@
 #include <linux/slab.h>
 #include <linux/dynamic_debug.h>
 #include <linux/libnvdimm.h>
+#include <linux/pfn_t.h>
 
 #include "drbd_int.h"
 #include "drbd_dax_pmem.h"
@@ -408,7 +409,7 @@ struct drbd_bitmap *drbd_bm_alloc(unsign
 {
 	struct drbd_bitmap *b;
 
-	b = kzalloc_obj(struct drbd_bitmap);
+	b = kzalloc(sizeof(struct drbd_bitmap), GFP_KERNEL);
 	if (!b)
 		return NULL;
 
@@ -1376,7 +1377,7 @@ static int bm_rw_range(struct drbd_devic
 	if (!expect(device, b->bm_number_of_pages))
 		return -ENODEV;
 
-	ctx = kmalloc_obj(struct drbd_bm_aio_ctx, GFP_NOIO);
+	ctx = kmalloc(sizeof(struct drbd_bm_aio_ctx), GFP_NOIO);
 	if (!ctx)
 		return -ENOMEM;
 
--- drbd_actlog.c
+++ /tmp/cocci-output-525084-573191-drbd_actlog.c
@@ -15,6 +15,7 @@
 #include <linux/crc32c.h>
 #include <linux/drbd.h>
 #include <linux/dynamic_debug.h>
+#include <linux/pfn_t.h>
 #include "drbd_int.h"
 #include "drbd_meta_data.h"
 #include "drbd_dax_pmem.h"
--- drbd-headers/linux/genl_magic_func-genl_register_family_with_ops_groups.h
+++ drbd-headers/linux/genl_magic_func-genl_register_family_with_ops_groups.h
@@ -18,8 +18,12 @@ static int CONCAT_(GENL_MAGIC_FAMILY, _genl_multicast_ ## group)(	\
 {									\
 	unsigned int group_id =						\
 		CONCAT_(GENL_MAGIC_FAMILY, _group_ ## group);		\
-	return genlmsg_multicast_allns(&ZZZ_genl_family, skb, 0,	\
-				 group_id);				\
+	int ret;		 					\
+	rcu_read_lock();						\
+	ret = genlmsg_multicast_allns(&ZZZ_genl_family, skb, 0,		\
+			 group_id, GFP_ATOMIC);				\
+	rcu_read_unlock();						\
+	return ret;							\
 }
 
 #include GENL_MAGIC_INCLUDE_FILE
